src/cunumeric/scan/scan_global_template.inl

```diff
@@ -0,0 +1,80 @@
+/* Copyright 2021-2022 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "cunumeric/scan/scan_global_util.h"
+#include "cunumeric/pitches.h"
+
+namespace cunumeric {
+
+using namespace Legion;
+using namespace legate;
+
+template <VariantKind KIND, ScanCode OP_CODE, LegateTypeCode CODE, int DIM>
+struct ScanGlobalImplBody;
+
+template <VariantKind KIND, ScanCode OP_CODE>
+struct ScanGlobalImpl {
+  template <LegateTypeCode CODE, int DIM>
+  void operator()(ScanGlobalArgs& args) const
+  {
+    using OP  = ScanOp<OP_CODE, CODE>;
+    using VAL = legate_type_of<CODE>;
+
+    auto out_rect      = args.out.shape<DIM>();
+    auto sum_vals_rect = args.sum_vals.shape<DIM>();
+
+    Pitches<DIM - 1> out_pitches;
+    size_t volume = out_pitches.flatten(out_rect);
+    Pitches<DIM - 1> sum_vals_pitches;
+    size_t sum_vals_volume = sum_vals_pitches.flatten(sum_vals_rect);
+
+    if (volume == 0) return;
+
+    auto out      = args.out.read_write_accessor<VAL, DIM>(out_rect);
+    auto sum_vals = args.sum_vals.read_accessor<VAL, DIM>(sum_vals_rect);
+
+    OP func;
+    ScanGlobalImplBody<KIND, OP_CODE, CODE, DIM>()(func,
+                                                   out,
+                                                   sum_vals,
+                                                   out_pitches,
+                                                   out_rect,
+                                                   sum_vals_pitches,
+                                                   sum_vals_rect,
+                                                   args.partition_index);
+  }
+};
+
+template <VariantKind KIND>
+struct ScanGlobalDispatch {
+  template <ScanCode OP_CODE>
+  void operator()(ScanGlobalArgs& args) const
+  {
+    return double_dispatch(args.out.dim(), args.out.code(), ScanGlobalImpl<KIND, OP_CODE>{}, args);
+  }
+};
+
+template <VariantKind KIND>
+static void scan_global_template(TaskContext& context)
+{
+  ScanGlobalArgs args{context.inputs()[1],
+                      context.outputs()[0],
+                      context.scalars()[0].value<ScanCode>(),
+                      context.get_task_index()};
```

Comment:
I suspect your compiler didn't warn you about this, but you shouldn't keep a reference to a temporary object in a struct. (note that `get_task_index` returns a fresh `DomainPoint` and not a reference.) if you want to pass it as a reference, please define the result in a stack first.

Reply:
e7c87896a7c50f0ec9d7b0f5f881a5910f51e20c